WordPress/gutenberg
# Parser: Client-side wpautop

Related: #3745, #2806

This pull request seeks to explore solutions for migrating classic content transparently, ~while also considering paragraph block serialization omitting comment demarcations~. In #3745, it was found that the server serialization behavior was insufficient to respect nested blocks. This was originally implemented to selectively apply wpautop to non-block content. The approach here instead assumes this to occur within Gutenberg when loading a post with non-block content. Simultaneously it adds parser awareness of the paragraph tag to upgrade to a paragraph block.

An advantage of implementing wpautop upgrading in the Gutenberg editor is that it allows the `the_content` filter for automatic paragraphs to be skipped altogether for posts containing blocks (already exists) and eliminates the need for server processing to this effect on save.

__Caveats:__

There are a few uncertainties with the changes proposed here:

- ~We don't want to call on the `wp.oldEditor.autop` global for applying autop, not just because it's a global, but because it is not functional in test environments. Instead, we should consider extracting these utilities (`autop`, `removep`) to a separate `@wordpress/autop` package.~ _**Edit:** Updated to use the new [`@wordpress/autop` package](https://www.npmjs.com/package/@wordpress/autop)_
- ~To this point, I had started the work here, and found in porting the PHP `wpautop` tests to JavaScript that there are many differences between the PHP and JavaScript implementations. We should seek to improve this.~ _**Edit:** Accounted for in the new [`@wordpress/autop` package](https://www.npmjs.com/package/@wordpress/autop)_
   - Related: https://core.trac.wordpress.org/ticket/42531#comment:23
- ~The undemarcated paragraph block is certainly less verbose, but I worry that this is a post-grammar-parse step to convert a freeform string including paragraph tags to a segmented set of freeform and paragraph blocks. This would also mean the server parser would not detect undemarcated paragraphs as paragraph blocks, even if within the Gutenberg editor they were treated as such.~ _**Edit:** No longer pursuing the undemarcated paragraphs as part of the changes here_

__Testing instructions:__

Verify that saving a post in the classic editor with removep'd content can be loaded into Gutenberg respecting paragraph separations. Example:

```
classic text

extra p

<div>foo</div>

more p
```

~This text, when loaded in Gutenberg, should produce the following sequence: 2 paragraph blocks, 1 freeform block (the `div`), and a last paragraph block. Saving the post should save with `<p>` tags.~ _**Edit:** No longer pursuing the undemarcated paragraphs as part of the changes here_

~Applying background color or alignment to a paragraph should apply comment demarcations, since these attributes are serialized in comment syntax.~_**Edit:** No longer pursuing the undemarcated paragraphs as part of the changes here_
  
Ensure that this content loads into the Gutenberg editor with paragraph blocks applied, and that even after inserting new block content and saving the post that it appears on the front-end with paragraphs still existing.
